# 字符串的基础检索与转换

In [1]:
import os; nb_dir = os.getcwd(); 
import sys; 
sys.path.append(
    os.sep.join([nb_dir, os.pardir, os.pardir, "Customized_Package"])
); 

In [2]:
import ipynb_quick_render

In [3]:
motto = "Verba volant, scripta manent"; 

In [4]:
#构造Unicode从U+0000到U+FFFF的所有字符顺次连接组成的字符串
char = str().join( 
    chr(x) for x in range(65536)
); 

## 用索引访问子串

### 用索引访问单个字符

In [5]:
print(motto)
print(motto[0], motto[3], motto[-1], motto[-6]); 

Verba volant, scripta manent
V b t m


### 用索引蔟按照一定顺序访问元素, 使得访问的结果构成子串
* `start: end: step`中的末项, 是`end`索引的元素的上一个元素, \
    因此索引簇索引的是"左闭右开区间". 

In [6]:
print(motto)
print(motto[6: 11], motto[0: 5: 1], motto[-6: -1]); 
print(motto[-6: ], motto[: 6: 1]); 

Verba volant, scripta manent
volan Verba manen
manent Verba 


## 字符串中位置和计数信息的查询

* 查询字符串的长度 (包含的字符总数): `len(s)`或`s.__len__()`
    * 由于字符编码的差异, 通常情况下, 字符串的**字符数少于其字节数**; 
    * 对于**采用变长编码规则**的字符串, 一般无法直接通过将字符数乘以某个整数的方式推断其字节数. 
* 查询一个子串`ch`在字符串中...
    * ...是否存在: `chr in s`或`s.__contains__(ch)`
    * ...首次(或末次)出现的位序: `s.index(ch)`, `s.find(ch)`
        或`s.rindex(ch)`, `s.rfind(ch)`
        * 当 `ch not in s`时, `index`和`rindex`方法将抛出`ValueError`, 
            而`find`和`rfind`方法将返回`-1`
    * ...出现的次数: `s.count(ch)`
    * ...是否位于其开头或者结尾: `s.startswith(ch)`或`s.endswith(ch)`

In [7]:
print(motto)
print(len(motto))
props = ("__contains__", "count", "index", "rindex", "startswith", "endswith"); 
vals = tuple(motto.__getattribute__(prop)("nt") for prop in props); 
ipynb_quick_render.Table2d(list(zip(props, vals))).render()

Verba volant, scripta manent
28


__contains__,True
count,2
index,10
rindex,26
startswith,False
endswith,True


## 字符串与元组的异同
字符串与元组均为不可变对象, 因此无法对字符串中的个别字符或子串进行修改, 只能通过拼接, 分割, 替换等操作构建新的字符串. 

值得注意的是: 字符串也可作为可迭代对象用于构建`generator`, 每次**无间隔地顺次取出其中的一个字符**; 字符串还支持**与元组类似的批量挂载和变长挂载**语法. 在部分程序的挂载语句和循环语句中需要格外注意. 

字符串与元组的主要差异体现在: 元组可以形成多级嵌套结构, 但字符串中不同子串之间的地位是平等的. 

In [8]:
#字符串作为可迭代对象参与遍历
for ch in "ABCD": 
    print(ch * 2, end="\x20")

AA BB CC DD 

In [9]:
#字符串的批量挂载
a, b = "AB"; print(a, b); 
a, b, a = "ABC"; print(a); 

A B
C


In [10]:
#字符串的变长挂载
res = list(); 
for str_mnt in "AB", "ABC", "ABCD": 
    a, *b, c = str_mnt; 
    res.append((str_mnt, a, b, c)); 
ipynb_quick_render.Table2d(res).render(); 

AB,A,[],B
ABC,A,['B'],C
ABCD,A,"['B', 'C']",D


In [11]:
#在字符串中, 无论索引多少次, 所得的结果仍为字符串
sub_motto = motto[6:12]; subsub_motto = sub_motto[0]; 
print(subsub_motto == motto[6]); 
print(type(subsub_motto) is type(motto))

True
True


## 字符类型判定

|函数|功能|备注|
|:-|:-:|:-|
|`s.isprintable()`|字符串中的所有字符是否<br>均为字符集中的可打印字符||
|`s.isalpha()`|字符串中的所有字符是否<br>均为特定语言文字中, <br>可以携带语义信息的字符|包括音位文字, 音节文字, 表意文字等|
|`s.isnumeric()`|字符串中的所有字符是否<br>均为表示数目的字符|除了`0`~`9`外, 包括各种语言文字中<br>使用的数字字符, 包括阿拉伯语数字, <br>梵语数字, 罗马数字(大小写), <br>汉字数字(大小写)等, 以及它们的<br>**各种变体**|
|`s.isdecimal()`|字符串中的所有字符是否<br>均为可用于在十进制下表示<br>**基数词**各位的数码的字符|**任何**一个字符应当可以用于<br>基数词的**任何一位**, 且任何相邻<br>两位之间无需其他字符作为分隔|
|`s.isalnum()`|字符串中的所有字符是否<br>均为携带语义信息或者<br>表示数目的字符||

* `isalpha`方法支持的字符, **包括但不限于**: 
    * 全音素文字, 如拉丁字母, 希腊字母, 西里尔字母等; 
    * 辅音音素文字, 如阿拉伯字母等; 
    * 元音附标文字, 如天城字母, 泰米尔字母, 孟加拉字母等; 
    * 音节文字, 如假名, 谚文, 彝文等; 
    * 表意文字, 如汉字等
* `isalpha`不支持**全角或半角**的标点符号; 不支持**被圆圈括注**的字母 (如`Ⓐ`)

* `isnumeric`方法支持的数字字符变体, **包括但不限于**: 
    * 使用括号或者圆圈包围的数字 (**只能用于序数词**, 因此对其调用`isdecimal`方法将返回`False`); 
    * `0`~`9`的上下标形式; 
    * 以**单字符形式**表示的常见分数 (约分至最简, **不能**用于构造`Fraction`对象); 
    * 汉字小写数字及部分数字的俗名, 如`亖`, `幺`等; 
    * 汉字大写数字及历史上通用的异体字, 如`参`, `漆`, `阡`, `陌`等
* `isnumeric`是`isdecimal`的**必要, 但非充分**条件. 

* `s.isalnum()`等价于
```python
s.isalpha() or s.isnumeric()
```
* 前述的汉字数字大小写使得`isalpha`与`isnumeric`同时成立, 因此`isalpha`与`isnumeric`条件不互斥

In [12]:
#构造生成器, 用于从可迭代对象中有重叠顺次获取相邻两个元素
import copy;  
def adjecent(iter_0): 
    iter_1 = copy.copy(iter_0).__iter__(); 
    iter_2 = copy.copy(iter_0).__iter__(); 
    iter_2.__next__(); 
    for elem_2 in iter_2: 
        elem_1 = iter_1.__next__(); 
        yield elem_1, elem_2; 

In [13]:
#计算一系列按升序排列的整数列表中, 相邻整数的起止范围
def integer_interval(iter_0): 
    intvl_ord_lower = [iter_0[0]]; 
    intvl_ord_upper = list(); 
    for ord_min, ord_max in adjecent(iter_0): 
        if ord_max - ord_min >= 2: 
            intvl_ord_lower.append(ord_max); 
            intvl_ord_upper.append(ord_min); 
    return zip(intvl_ord_lower, intvl_ord_upper); 

In [14]:
#控制字符的个数
str_ctrl = str().join(
    [ch for (x, ch) in enumerate(char) if not ch.isprintable()]
); 
len(str_ctrl)

10249

In [15]:
#计算所有控制字符的码点分布范围
str_ctrl_ord = list(ord(ch) for ch in str_ctrl); 
ipynb_quick_render.Table2d(
    list(
        ("U+{start:0>4x}~U+{end:0>4x}".format(start=start, end=end), )
         for start, end in integer_interval(str_ctrl_ord)
    ), multi_column=6
).render(); 

U+0000~U+001f,U+0a92~U+0a92,U+0cd7~U+0cdd,U+12b1~U+12b1,U+1f58~U+1f58,U+a62c~U+a63f
U+007f~U+00a0,U+0aa9~U+0aa9,U+0cdf~U+0cdf,U+12b6~U+12b7,U+1f5a~U+1f5a,U+a6f8~U+a6ff
U+00ad~U+00ad,U+0ab1~U+0ab1,U+0ce4~U+0ce5,U+12bf~U+12bf,U+1f5c~U+1f5c,U+a7ba~U+a7f6
U+0378~U+0379,U+0ab4~U+0ab4,U+0cf0~U+0cf0,U+12c1~U+12c1,U+1f5e~U+1f5e,U+a82c~U+a82f
U+0380~U+0383,U+0aba~U+0abb,U+0cf3~U+0cff,U+12c6~U+12c7,U+1f7e~U+1f7f,U+a83a~U+a83f
U+038b~U+038b,U+0ac6~U+0ac6,U+0d04~U+0d04,U+12d7~U+12d7,U+1fb5~U+1fb5,U+a878~U+a87f
U+038d~U+038d,U+0aca~U+0aca,U+0d0d~U+0d0d,U+1311~U+1311,U+1fc5~U+1fc5,U+a8c6~U+a8cd
U+03a2~U+03a2,U+0ace~U+0acf,U+0d11~U+0d11,U+1316~U+1317,U+1fd4~U+1fd5,U+a8da~U+a8df
U+0530~U+0530,U+0ad1~U+0adf,U+0d45~U+0d45,U+135b~U+135c,U+1fdc~U+1fdc,U+a954~U+a95e
U+0557~U+0558,U+0ae4~U+0ae5,U+0d49~U+0d49,U+137d~U+137f,U+1ff0~U+1ff1,U+a97d~U+a97f
U+058b~U+058c,U+0af2~U+0af8,U+0d50~U+0d53,U+139a~U+139f,U+1ff5~U+1ff5,U+a9ce~U+a9ce


In [16]:
#文本字符的个数
str_alpha = str().join(
    [ch for (x, ch) in enumerate(char) if ch.isalpha()]
); 
len(str_alpha)

48832

In [17]:
#计算所有文本字符的码点分布范围
str_alpha_ord = list(ord(ch) for ch in str_alpha); 
ipynb_quick_render.Table2d(
    list(
        ("U+{start:0>4x}~U+{end:0>4x}".format(start=start, end=end), )
         for start, end in integer_interval(str_alpha_ord)
    ), multi_column=6
).render(); 

U+0041~U+005a,U+09dc~U+09dd,U+0d12~U+0d3a,U+1380~U+138f,U+2071~U+2071,U+a80c~U+a822
U+0061~U+007a,U+09df~U+09e1,U+0d3d~U+0d3d,U+13a0~U+13f5,U+207f~U+207f,U+a840~U+a873
U+00aa~U+00aa,U+09f0~U+09f1,U+0d4e~U+0d4e,U+13f8~U+13fd,U+2090~U+209c,U+a882~U+a8b3
U+00b5~U+00b5,U+09fc~U+09fc,U+0d54~U+0d56,U+1401~U+166c,U+2102~U+2102,U+a8f2~U+a8f7
U+00ba~U+00ba,U+0a05~U+0a0a,U+0d5f~U+0d61,U+166f~U+167f,U+2107~U+2107,U+a8fb~U+a8fb
U+00c0~U+00d6,U+0a0f~U+0a10,U+0d7a~U+0d7f,U+1681~U+169a,U+210a~U+2113,U+a8fd~U+a8fe
U+00d8~U+00f6,U+0a13~U+0a28,U+0d85~U+0d96,U+16a0~U+16ea,U+2115~U+2115,U+a90a~U+a925
U+00f8~U+02c1,U+0a2a~U+0a30,U+0d9a~U+0db1,U+16f1~U+16f8,U+2119~U+211d,U+a930~U+a946
U+02c6~U+02d1,U+0a32~U+0a33,U+0db3~U+0dbb,U+1700~U+170c,U+2124~U+2124,U+a960~U+a97c
U+02e0~U+02e4,U+0a35~U+0a36,U+0dbd~U+0dbd,U+170e~U+1711,U+2126~U+2126,U+a984~U+a9b2
U+02ec~U+02ec,U+0a38~U+0a39,U+0dc0~U+0dc6,U+1720~U+1731,U+2128~U+2128,U+a9cf~U+a9cf


In [18]:
#表示数目的字符
str_num = str().join(
    [ch for (x, ch) in enumerate(char) if ch.isnumeric()]
); 
str_num

'0123456789²³¹¼½¾٠١٢٣٤٥٦٧٨٩۰۱۲۳۴۵۶۷۸۹߀߁߂߃߄߅߆߇߈߉०१२३४५६७८९০১২৩৪৫৬৭৮৯৴৵৶৷৸৹੦੧੨੩੪੫੬੭੮੯૦૧૨૩૪૫૬૭૮૯୦୧୨୩୪୫୬୭୮୯୲୳୴୵୶୷௦௧௨௩௪௫௬௭௮௯௰௱௲౦౧౨౩౪౫౬౭౮౯౸౹౺౻౼౽౾೦೧೨೩೪೫೬೭೮೯൘൙൚൛൜൝൞൦൧൨൩൪൫൬൭൮൯൰൱൲൳൴൵൶൷൸෦෧෨෩෪෫෬෭෮෯๐๑๒๓๔๕๖๗๘๙໐໑໒໓໔໕໖໗໘໙༠༡༢༣༤༥༦༧༨༩༪༫༬༭༮༯༰༱༲༳၀၁၂၃၄၅၆၇၈၉႐႑႒႓႔႕႖႗႘႙፩፪፫፬፭፮፯፰፱፲፳፴፵፶፷፸፹፺፻፼ᛮᛯᛰ០១២៣៤៥៦៧៨៩៰៱៲៳៴៵៶៷៸៹᠐᠑᠒᠓᠔᠕᠖᠗᠘᠙᥆᥇᥈᥉᥊᥋᥌᥍᥎᥏᧐᧑᧒᧓᧔᧕᧖᧗᧘᧙᧚᪀᪁᪂᪃᪄᪅᪆᪇᪈᪉᪐᪑᪒᪓᪔᪕᪖᪗᪘᪙᭐᭑᭒᭓᭔᭕᭖᭗᭘᭙᮰᮱᮲᮳᮴᮵᮶᮷᮸᮹᱀᱁᱂᱃᱄᱅᱆᱇᱈᱉᱐᱑᱒᱓᱔᱕᱖᱗᱘᱙⁰⁴⁵⁶⁷⁸⁹₀₁₂₃₄₅₆₇₈₉⅐⅑⅒⅓⅔⅕⅖⅗⅘⅙⅚⅛⅜⅝⅞⅟ⅠⅡⅢⅣⅤⅥⅦⅧⅨⅩⅪⅫⅬⅭⅮⅯⅰⅱⅲⅳⅴⅵⅶⅷⅸⅹⅺⅻⅼⅽⅾⅿↀↁↂↅↆↇↈ↉①②③④⑤⑥⑦⑧⑨⑩⑪⑫⑬⑭⑮⑯⑰⑱⑲⑳⑴⑵⑶⑷⑸⑹⑺⑻⑼⑽⑾⑿⒀⒁⒂⒃⒄⒅⒆⒇⒈⒉⒊⒋⒌⒍⒎⒏⒐⒑⒒⒓⒔⒕⒖⒗⒘⒙⒚⒛⓪⓫⓬⓭⓮⓯⓰⓱⓲⓳⓴⓵⓶⓷⓸⓹⓺⓻⓼⓽⓾⓿❶❷❸❹❺❻❼❽❾❿➀➁➂➃➄➅➆➇➈➉➊➋➌➍➎➏➐➑➒➓⳽〇〡〢〣〤〥〦〧〨〩〸〹〺㆒㆓㆔㆕㈠㈡㈢㈣㈤㈥㈦㈧㈨㈩㉈㉉㉊㉋㉌㉍㉎㉏㉑㉒㉓㉔㉕㉖㉗㉘㉙㉚㉛㉜㉝㉞㉟㊀㊁㊂㊃㊄㊅㊆㊇㊈㊉㊱㊲㊳㊴㊵㊶㊷㊸㊹㊺㊻㊼㊽㊾㊿㐅㒃㠪㭍一七万三九二五亖亿什仟仨伍佰億兆兩八六十千卄卅卌叁参參叄四壱壹幺廾廿弌弍弎弐拾捌柒漆玖百肆萬貮貳贰阡陆陌陸零꘠꘡꘢꘣꘤꘥꘦꘧꘨꘩ꛦꛧꛨꛩꛪꛫꛬꛭꛮꛯ꠰꠱꠲꠳꠴꠵꣐꣑꣒꣓꣔꣕꣖꣗꣘꣙꤀꤁꤂꤃꤄꤅꤆꤇꤈꤉꧐꧑꧒꧓꧔꧕꧖꧗꧘꧙꧰꧱꧲꧳꧴꧵꧶꧷꧸꧹꩐꩑꩒꩓꩔꩕꩖꩗꩘꩙꯰꯱꯲꯳꯴꯵꯶꯷꯸꯹參拾兩零六陸什０１２３４５６７８９'

In [19]:
#可用于十进制数码的字符
str_dec = str().join(
    [ch for (x, ch) in enumerate(char) if ch.isdecimal()]
); 
str_dec

'0123456789٠١٢٣٤٥٦٧٨٩۰۱۲۳۴۵۶۷۸۹߀߁߂߃߄߅߆߇߈߉०१२३४५६७८९০১২৩৪৫৬৭৮৯੦੧੨੩੪੫੬੭੮੯૦૧૨૩૪૫૬૭૮૯୦୧୨୩୪୫୬୭୮୯௦௧௨௩௪௫௬௭௮௯౦౧౨౩౪౫౬౭౮౯೦೧೨೩೪೫೬೭೮೯൦൧൨൩൪൫൬൭൮൯෦෧෨෩෪෫෬෭෮෯๐๑๒๓๔๕๖๗๘๙໐໑໒໓໔໕໖໗໘໙༠༡༢༣༤༥༦༧༨༩၀၁၂၃၄၅၆၇၈၉႐႑႒႓႔႕႖႗႘႙០១២៣៤៥៦៧៨៩᠐᠑᠒᠓᠔᠕᠖᠗᠘᠙᥆᥇᥈᥉᥊᥋᥌᥍᥎᥏᧐᧑᧒᧓᧔᧕᧖᧗᧘᧙᪀᪁᪂᪃᪄᪅᪆᪇᪈᪉᪐᪑᪒᪓᪔᪕᪖᪗᪘᪙᭐᭑᭒᭓᭔᭕᭖᭗᭘᭙᮰᮱᮲᮳᮴᮵᮶᮷᮸᮹᱀᱁᱂᱃᱄᱅᱆᱇᱈᱉᱐᱑᱒᱓᱔᱕᱖᱗᱘᱙꘠꘡꘢꘣꘤꘥꘦꘧꘨꘩꣐꣑꣒꣓꣔꣕꣖꣗꣘꣙꤀꤁꤂꤃꤄꤅꤆꤇꤈꤉꧐꧑꧒꧓꧔꧕꧖꧗꧘꧙꧰꧱꧲꧳꧴꧵꧶꧷꧸꧹꩐꩑꩒꩓꩔꩕꩖꩗꩘꩙꯰꯱꯲꯳꯴꯵꯶꯷꯸꯹０１２３４５６７８９'

In [20]:
#表示文本或者数目的字符
str_alnum = str().join(
    [ch for (x, ch) in enumerate(char) if ch.isalnum()]
); 

In [21]:
#isdecimal比isnumeric条件更严格, 是isnumeric的特例
set(str_num).issuperset(set(str_dec))

True

In [22]:
#字母字符集与数字字符集取并
set(str_num).union(set(str_alpha)) == set(str_alnum)

True

In [23]:
#文本字符集与数字字符集取交, 结果非空
str().join(set(str_num).intersection(set(str_alpha)))

'千玖参什肆廿六一億伍贰㒃万貳五卌叁仨參㐅亖零㭍拾陌兩參八壹仟阡七兆二弐十陆捌柒卅佰兩卄零拾陸弍弎廾陸九壱㠪六貮弌漆幺萬什百亿四三叄'

## 字符串的转换

### 字母的格式转换

此处所述的字母, 包括拉丁字母, 希腊字母, 西里尔字母, 格鲁吉亚字母等区分大小写, 且用于**标记音位的字母**, 以及它们的声调形式和主要变体. 

以下格式转换函数仅适用于上述字母的大小写转换, 无法实现汉字的简体-繁体转换, 日文的平假名-片假名转换等功能, 当函数遇到不区分大小写的字符时, 对这些字符不作处理. 

|函数|功能|备注|
|:-|:-:|:-|
|`s.upper()`|将字符串中的所有字母转换为大写||
|`s.lower()`|将字符串中的所有字母转换为小写||
|`s.swapcase()`|将字符串中的所有大写字母转换为<br>小写, 小写字母转换为大写||
|`s.capitalize()`|将字符串中, 出现在字符串开头的<br>第一个字母转换为大写, <br>其余位置的字母转换为小写||
|`s.title()`|将字符串中的所有单词中, <br>出现在每个单词开头的<br>第一个字母转换为大写, <br>其余位置的字母转换为小写|单词是由**连续的拼音<br>文字字母**构成的子串; <br>不能识别单词的词性|
|`s.isupper()`|字符串中包含的所有字母<br>是否均为大写|字母以外的字符数量和<br>位置不影响结果; <br>如果字符串中不包含上述<br>任何字母, 则返回`False`|
|`s.islower()`|字符串中包含的所有字母<br>是否均为小写|字母以外的字符数量和<br>位置不影响结果; <br>如果字符串中不包含上述<br>任何字母, 则返回`False`|

注意事项: 
* 部分连笔字母或者变调字母的小写形式中, 多个字母 (或字母与符号的组合) 被合成一个字符, 使用`upper`方法将该字符转化为大写形式后, 连写的字母 (或字母与符号的组合) 将被**拆分为对应的, 连续的多个大写字母** (或大写字母与符号的组合). 常见的案例包括:  
    * 小写字母`ß` (`U+00df`: 小写拉丁字母eszett) 是古字母`ſ` (`U+017f`: 小写拉丁字母长s) 和 `z` 的二合字母 (一说是`ſ`与`s`的二合字母)$^{[1–2]}$, 由于该字母不会用于德语单词的词首, 因此长期没有对应的大写字母, 其大写形式按照约定俗成, 采用其转写形式`ss`的大写, 即`SS`$^{[2]}$. 进入信息时代后, 在Unicode的发展过程中, 虽然新设计了该字母的大写形式`ẞ` (`U+1e9e`), 但出于兼容性方面的考虑, `ß` 的大写形式仍为`SS`, 而`ẞ`的小写形式为`ß`. 
    * 在部分印刷作品中, 小写字母组合`ff`, `fi`, `fl`, `ffi`, `ffl`分别连写成`ff`, `fi`, `fl`, `ffi`, `ffl` (`U+fb00`~`U+fb04`), 仅使用一个字符$^{[1‚3–5]}$, 但对应的大写形式不连写, 占用多个字符. 
* 部分字母的大写 (或小写) 形式被**分配了多个码点**, 分别用于表示不同的含义, 它们的小写 (或大写) 形式**采用同一字符同一码点**. 常见的案例包括:  
    * 小写字母`μ`占据了两个码点 (`U+00b5`: 公制单位前缀"百万分之一", `U+03bc`: 小写希腊字母mu)$^{[6‚7]}$, 它们的大写形式均为`Μ` (`U+039c`: 大写希腊字母Mu); 而`Μ`使用`lower`方法所得的小写形式为`U+03bc`; 
    * 小写字母`σ` (`U+03c3`: 小写希腊字母sigma) 和小写字母`ς` (`U+03c2`: 专用于词尾的小写希腊字母sigma) 的大写形式均为`Σ`(`U+03a3`: 大写希腊字母Sigma)$^{[1‚8‚9]}$; 而`Σ`使用`lower`方法所得的小写形式为`U+03c3`; 
        * 与之类似的还有`θ`和`ϑ`, `φ`和`ϕ`, `π`和`ϖ` (后者专用于词首) , `ρ`和`ϱ` (后者专用于斜体) 等$^{[7]}$
        * 上述数对希腊字母中, 每对字母使用`upper`方法得到的大写字母, 均为同一字符同一码点, 但大写字母使用`lower`方法转化为小写形式时, 均采用字母对中的前一个字母
    * 大写字母`Å`占据了两个码点 (`U+00c5`: 大写拉丁字母A上方带圈, `U+212b`: 长度单位Ångstrom)$^{[9]}$, 它们的小写形式均为`å`(`U+00e5`: 小写拉丁字母a上方带圈); 而`å`使用`upper`方法所得的大写形式为`U+00c5`. 
        * 与之类似的还有表示电阻单位Ohm的`Ω`(`U+2126`)$^{[7]}$等其他兼用于计量单位符号的拉丁和希腊字母
    
在上述情况下, 大写字母和小写字母并非一一映射的关系, 因此`upper`方法和`lower`方法在遇到包含上述字母的字符串时, 不是互逆操作. 在进行部分语言的本地化适配, 或者处理部分科技类文章使用的符号时, 需要格外注意上述问题. 

参考文献: 
```
[1] Haralambous Y, Plaice J, Braams J. Never again active characters! 
    Ω-Babel[J]. TUGboat, 1995, 16(4): 418–427.

[2] Pitchford J. Dutch, German, Austrian, Flemish and Afrikaans names[J]. 
    The Indexer: The International Journal of Indexing, 2006, 25(2): C11–C14.

[3] Lubran A. Word Containing the FFL Trigram[J]. Word Ways, 1991, 24(2): 16.

[4] Hoenig A. Virtual fonts, virtuous fonts[J]. TUGboat, 1997, 18(2): 113–121.

[5] Rahma AMS, Bhaya WS, Al-Nasrawi DA. Data hiding method for english scripts 
    using ligature characters unicode[J]. 
    European Journal of Scientific Research, 2013, 112(4): 452–459.

[6] Lu CJ, Browne AC. Converting unicode lexicon and lexical tools for ASCII 
    NLP applications[C]//AMIA Annu Symp Proc. 2011, 2011: 1870.

[7] Foster MP. Principles for constructing notation in unit systems and 
    their application to the SI[J]. Accreditation and Quality Assurance, 
    2012, 17(1): 85–92.

[8] Haralambous Y. From Unicode to typography, a case study: 
    the Greek script[C]//Fourteenth International Unicode Conference. 
    1998: b–10.

[9] Moran S, Cysouw M. The Unicode Cookbook for Linguists: Managing writing 
    systems using orthography profiles[M]. Berlin: Language Science Press, 
    2017, 10.
```

In [24]:
#区分大小写形式的字符
str_case_diff = str().join(
    [ch for (x, ch) in enumerate(char) if ch.isupper() or ch.islower()]
); 
str_case_diff

'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzªµºÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿĀāĂ㥹ĆćĈĉĊċČčĎďĐđĒēĔĕĖėĘęĚěĜĝĞğĠġĢģĤĥĦħĨĩĪīĬĭĮįİıIJijĴĵĶķĸĹĺĻļĽľĿŀŁłŃńŅņŇňʼnŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚśŜŝŞşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſƀƁƂƃƄƅƆƇƈƉƊƋƌƍƎƏƐƑƒƓƔƕƖƗƘƙƚƛƜƝƞƟƠơƢƣƤƥƦƧƨƩƪƫƬƭƮƯưƱƲƳƴƵƶƷƸƹƺƼƽƾƿDŽdžLJljNJnjǍǎǏǐǑǒǓǔǕǖǗǘǙǚǛǜǝǞǟǠǡǢǣǤǥǦǧǨǩǪǫǬǭǮǯǰDZdzǴǵǶǷǸǹǺǻǼǽǾǿȀȁȂȃȄȅȆȇȈȉȊȋȌȍȎȏȐȑȒȓȔȕȖȗȘșȚțȜȝȞȟȠȡȢȣȤȥȦȧȨȩȪȫȬȭȮȯȰȱȲȳȴȵȶȷȸȹȺȻȼȽȾȿɀɁɂɃɄɅɆɇɈɉɊɋɌɍɎɏɐɑɒɓɔɕɖɗɘəɚɛɜɝɞɟɠɡɢɣɤɥɦɧɨɩɪɫɬɭɮɯɰɱɲɳɴɵɶɷɸɹɺɻɼɽɾɿʀʁʂʃʄʅʆʇʈʉʊʋʌʍʎʏʐʑʒʓʕʖʗʘʙʚʛʜʝʞʟʠʡʢʣʤʥʦʧʨʩʪʫʬʭʮʯʰʱʲʳʴʵʶʷʸˀˁˠˡˢˣˤͅͰͱͲͳͶͷͺͻͼͽͿΆΈΉΊΌΎΏΐΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩΪΫάέήίΰαβγδεζηθικλμνξοπρςστυφχψωϊϋόύώϏϐϑϒϓϔϕϖϗϘϙϚϛϜϝϞϟϠϡϢϣϤϥϦϧϨϩϪϫϬϭϮϯϰϱϲϳϴϵϷϸϹϺϻϼϽϾϿЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏАБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдежзийклмнопрстуфхцчшщъыьэюяѐёђѓєѕіїјљњћќѝўџѠѡѢѣѤѥѦѧѨѩѪѫѬѭѮѯѰѱѲѳѴѵѶѷѸѹѺѻѼѽѾѿҀҁҊҋҌҍҎҏҐґҒғҔҕҖҗҘҙҚқҜҝҞҟҠҡҢңҤҥҦҧҨҩҪҫҬҭҮүҰұҲҳҴҵҶҷҸҹҺһҼҽҾҿӀӁӂӃӄӅӆӇӈӉӊӋӌӍӎӏӐӑӒӓӔӕӖӗӘәӚӛӜӝӞӟӠӡӢӣӤӥӦӧӨөӪӫӬӭӮӯӰӱӲӳӴӵӶӷӸӹӺӻӼӽӾӿԀԁԂԃԄԅԆԇԈԉԊԋԌԍԎԏԐԑԒԓԔԕԖԗԘԙԚԛԜԝԞԟԠԡԢԣԤԥԦԧԨԩԪԫԬԭԮԯԱԲԳԴԵԶԷԸԹԺԻԼԽԾԿՀՁՂՃ

In [25]:
#大小写转换不互逆的小写字母
abnormal_lower = list(
    (hex(x), ch, ch.upper()) for (x, ch) in enumerate(char) 
    if ch.islower() and ch != ch.upper().lower()
); 
ipynb_quick_render.Table2d(abnormal_lower, multi_column=5).render(); 

0xb5,µ,Μ,0x1c82,ᲂ,О,0x1f83,ᾃ,ἋΙ,0x1fa7,ᾧ,ὯΙ,0x1fe7,ῧ,Ϋ͂
0xdf,ß,SS,0x1c83,ᲃ,С,0x1f84,ᾄ,ἌΙ,0x1fb2,ᾲ,ᾺΙ,0x1ff2,ῲ,ῺΙ
0x131,ı,I,0x1c84,ᲄ,Т,0x1f85,ᾅ,ἍΙ,0x1fb3,ᾳ,ΑΙ,0x1ff3,ῳ,ΩΙ
0x149,ʼn,ʼN,0x1c85,ᲅ,Т,0x1f86,ᾆ,ἎΙ,0x1fb4,ᾴ,ΆΙ,0x1ff4,ῴ,ΏΙ
0x17f,ſ,S,0x1c86,ᲆ,Ъ,0x1f87,ᾇ,ἏΙ,0x1fb6,ᾶ,Α͂,0x1ff6,ῶ,Ω͂
0x1f0,ǰ,J̌,0x1c87,ᲇ,Ѣ,0x1f90,ᾐ,ἨΙ,0x1fb7,ᾷ,Α͂Ι,0x1ff7,ῷ,Ω͂Ι
0x345,ͅ,Ι,0x1c88,ᲈ,Ꙋ,0x1f91,ᾑ,ἩΙ,0x1fbe,ι,Ι,0xfb00,ff,FF
0x390,ΐ,Ϊ́,0x1e96,ẖ,H̱,0x1f92,ᾒ,ἪΙ,0x1fc2,ῂ,ῊΙ,0xfb01,fi,FI
0x3b0,ΰ,Ϋ́,0x1e97,ẗ,T̈,0x1f93,ᾓ,ἫΙ,0x1fc3,ῃ,ΗΙ,0xfb02,fl,FL
0x3c2,ς,Σ,0x1e98,ẘ,W̊,0x1f94,ᾔ,ἬΙ,0x1fc4,ῄ,ΉΙ,0xfb03,ffi,FFI
0x3d0,ϐ,Β,0x1e99,ẙ,Y̊,0x1f95,ᾕ,ἭΙ,0x1fc6,ῆ,Η͂,0xfb04,ffl,FFL


In [26]:
#大小写转换不互逆的大写字母
abnormal_upper = list(
    (hex(x), ch, ch.lower()) for (x, ch) in enumerate(char) 
    if ch.isupper() and ch != ch.lower().upper()
); 
ipynb_quick_render.Table2d(abnormal_upper).render(); 

0x130,İ,i̇
0x3f4,ϴ,θ
0x1e9e,ẞ,ß
0x2126,Ω,ω
0x212a,K,k
0x212b,Å,å


In [27]:
#比较不同大小写转换函数的结果差异
#此处暂不将比对结果用ipynb_quick_render渲染成HTML表格
#因为Jupyter Notebook输出单元对HTML表格内容默认使用的字体是非等宽的, 不便对照
props = ("__str__", "upper", "lower", "swapcase", "capitalize", "title"); 
vals = tuple(motto.__getattribute__(prop)() for prop in props); 
for prop, val in zip(props, vals): 
    print("{prop:<12s} {val:<s}".format(prop=prop, val=val))

__str__      Verba volant, scripta manent
upper        VERBA VOLANT, SCRIPTA MANENT
lower        verba volant, scripta manent
swapcase     vERBA VOLANT, SCRIPTA MANENT
capitalize   Verba volant, scripta manent
title        Verba Volant, Scripta Manent
